### PR TITLE
HEEDLS-511 Delegate groups - edit group description

### PR DESCRIPTION
**Pull request by showkath10**
### JIRA link
https://softwiretech.atlassian.net/browse/HEEDLS-511

### Description
Added new page for user to be able to edit delegate group description.

### Screenshots
image
image

-----
### Developer checks
I have:
- [ x] Run the formatter and made sure there are no IDE errors.
- [ x] Written tests for the changes (controller, data services, services, view models etc) and manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [ ] Updated/added documentation in Swiki and/or Readme.
- [x] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [x] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

File: DigitalLearningSolutions.Data/DataServices/GroupsDataService.cs [hunk 1]
```diff
@@ -1,222 +1,250 @@
...
+                @"SELECT
+                        GroupLabel,
+                        GroupDescription
+                    FROM Groups
+                    WHERE GroupID = @groupId AND CentreId = @centreId",
+                new { groupId, centreId }
+            ).SingleOrDefault();
+        }
+
+        public bool UpdateGroupDescription(int groupId, int centreId, string groupDescription)
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Data/DataServices/GroupsDataService.cs:238]: The name of this isn't clear enough as to what is happening, we need to name it so that it is obvious that this is expected to sometimes fail and not actually update the database. Maybe Try..., or UpdateGroupDescriptionAndReturnIfSuccessful or something. The name of the unit tests then need updating to match.

**Review by AlexJacksonDS — CHANGES_REQUESTED** [review]
Quite a few things that need changing.

As well as all the below comments there are also a few extra things I couldn't sensibly comment against the files.

The GroupsDataServiceTests and GroupsDataService have had their line endings changed from CRLF to LF, can you change these back (let me know if you don't know how). Can you check that `git config core.autocrlf` returns false, and if not set this with `git config --global core.autocrlf false`. This should prevent line endings changes happening unexpectedly, but if you do come across any files with mixed line ending, then CRLF is the ending we should be using. (We think some of the merge tools could also have been messing this up in the past)

There are no automated accessibility tests for the new page. In this case it should be simple, adding a new InlineData in BasicAuthenticatedAccessibilityTests using a suitable groupId for the centre.

There is also a missed requirement on the ticket, setting the maximum length of the Description field to 1000 characters, including creating a migration to do this on the database. I can see that perhaps this might not be desired any more, for example if live has a bunch of descriptions over 1000 characters. I don't know if there were any discussions with @SteveJacksonSoft / @stellake about this requirement, but either this is missing from the pull request, or the task to update the Jira in the MR checklist hasn't been completed.

File: DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/DelegateGroups/EditDelegateGroupDescription.cshtml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+  ViewData["Application"] = "Tracking System";
+  ViewData["HeaderPath"] = $"{Configuration["AppRootPath"]}/TrackingSystem/Centre/Dashboard";
+  ViewData["HeaderPathName"] = "Tracking System";
+}
+
+<link rel="stylesheet" href="@Url.Content("~/css/trackingSystem/delegates/groupDelegates.css")" asp-append-version="true">
+
+@section NavMenuItems {
+  <partial name="~/Views/TrackingSystem/Shared/_NavMenuItems.cshtml" />
+}
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/DelegateGroups/EditDelegateGroupDescription.cshtml:17]: Extra thing I didn't know about until the review on one of my tickets, can you use the _AutoNavMenuItems partial instead.
`<partial name="_AutoNavMenuItems" model="ApplicationType.TrackingSystem"/>`

File: DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/DelegateGroupsController.cs [hunk 3]
```diff
@@ -179,6 +179,38 @@ public IActionResult GroupCourses(int groupId, int page = 1)
...
         }
 
+        [Route("{groupId:int}/EditDelegateGroupDescription")]
+        [HttpGet]
+        public IActionResult EditDelegateGroupDescription(int groupId)
+        {
+            var centreId = User.GetCentreId();
+            var group = groupsDataService.GetGroup(groupId, centreId);
+
+            if (group is null)
```
[the remark this answers is not in the document]
Comment (showkath10) [reply DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/DelegateGroupsController.cs]: updated with ==

File: DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/DelegateGroupsController.cs [hunk 4]
```diff
@@ -179,6 +179,38 @@ public IActionResult GroupCourses(int groupId, int page = 1)
             return View(model);
         }
 
+        [Route("{groupId:int}/EditDelegateGroupDescription")]
```
[the remark this answers is not in the document]
Comment (showkath10) [reply DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/DelegateGroupsController.cs]: Action method  name is shortened and now it is EditDescription as suggested

File: DigitalLearningSolutions.Web/ViewModels/TrackingSystem/Delegates/DelegateGroups/EditDelegateGroupDescriptionViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,20 @@
+﻿namespace DigitalLearningSolutions.Web.ViewModels.TrackingSystem.Delegates.DelegateGroups
+{
+    using DigitalLearningSolutions.Data.Models.DelegateGroups;
+
+    public class EditDelegateGroupDescriptionViewModel
+    {
+        public int GroupId { get; set; }
+        public string? GroupLabel { get; set; }
```
[the remark this answers is not in the document]
Comment (showkath10) [reply DigitalLearningSolutions.Web/ViewModels/TrackingSystem/Delegates/DelegateGroups/EditDelegateGroupDescriptionViewModel.cs]: As AdminUser class has property GroupLabel so I used it for consistency. I have now changed it to GroupName as suggested.

File: DigitalLearningSolutions.Web/ViewModels/TrackingSystem/Delegates/DelegateGroups/EditDelegateGroupDescriptionViewModel.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+        public string? Description { get; set; }
```
[the remark this answers is not in the document]
Comment (showkath10) [reply DigitalLearningSolutions.Web/ViewModels/TrackingSystem/Delegates/DelegateGroups/EditDelegateGroupDescriptionViewModel.cs]: Sorry, I missed this one. I have added this validation now.

File: DigitalLearningSolutions.Web/ViewModels/TrackingSystem/Delegates/DelegateGroups/EditDelegateGroupDescriptionViewModel.cs [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+        public EditDelegateGroupDescriptionViewModel() { }
```
[the remark this answers is not in the document]
Comment (showkath10) [reply DigitalLearningSolutions.Web/ViewModels/TrackingSystem/Delegates/DelegateGroups/EditDelegateGroupDescriptionViewModel.cs:9]: Sorry, I ran the formatters and it is formatted correctly inline other files.

File: DigitalLearningSolutions.Data/DataServices/GroupsDataService.cs [hunk 8]
```diff
@@ -1,222 +1,250 @@
...
+        {
+            return connection.Query<Group>(
+                @"SELECT
+                        GroupLabel,
+                        GroupDescription
+                    FROM Groups
+                    WHERE GroupID = @groupId AND CentreId = @centreId",
+                new { groupId, centreId }
+            ).SingleOrDefault();
+        }
```
[the remark this answers is not in the document]
Comment (showkath10) [reply DigitalLearningSolutions.Data/DataServices/GroupsDataService.cs]: I have combined GetGroupName and GetGroup into new method GetGroupAtCentreById and removed GetGroupName and GetGroup methods. This method now populates available properties.

File: DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs [hunk 9]
```diff
@@ -1,294 +1,326 @@
...
+            const int centerId = 101;
+            const int groupId = 5;
+            const string expectedDescription = "Test group description";
+
+            // When
+            groupsDataService.UpdateGroupDescription(groupId, centerId, expectedDescription);
+
+            //Then
+            var result = groupsDataService.GetGroup(groupId, centerId);
+            result?.GroupDescription.Should().Be(expectedDescription);
```
[the remark this answers is not in the document]
Comment (showkath10) [reply DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs]: Sorry, yes, it should have been in transaction. I updated the code.

File: DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs [hunk 10]
```diff
@@ -1,294 +1,326 @@
...
+        }
+
+        [Test]
+        public void UpdateGroupDescription_with_incorrect_centreId_does_not_update_record()
+        {
+            // Given
+            const int centerId = 107;   //Incorrect centre id
```
[the remark this answers is not in the document]
Comment (showkath10) [reply DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs]: I have updated variable name as suggested

File: DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs [hunk 11]
```diff
@@ -1,294 +1,326 @@
...
+            const int groupId = 5;
+            const string expectedDescription = "Test group description";
+
+            // When
+            groupsDataService.UpdateGroupDescription(groupId, centerId, expectedDescription);
+
+            //Then
+            var result = groupsDataService.GetGroup(groupId, centerId);
+            result?.GroupDescription.Should().NotBe(expectedDescription);
```
[the remark this answers is not in the document]
Comment (showkath10) [reply DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs]: Thanks, it is now moved to single method GetGroupAtCentreById as suggested

File: DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/DelegateGroups/EditDelegateGroupDescription.cshtml (hunk 2 above)
Comment (showkath10) [reply DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/DelegateGroups/EditDelegateGroupDescription.cshtml]: Sorry, I didn't know this. I updated the code to as above partial view

File: DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/DelegateGroups/EditDelegateGroupDescription.cshtml [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+
+<div class="nhsuk-grid-row">
+  <div class="nhsuk-grid-column-full">
+    <h1 class="nhsuk-heading-l">@ViewData["Title"]</h1>
```
[the remark this answers is not in the document]
Comment (showkath10) [reply DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/DelegateGroups/EditDelegateGroupDescription.cshtml]: I have updated h1 with -xl as suggested

File: DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/DelegateGroups/EditDelegateGroupDescription.cshtml [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+@model EditDelegateGroupDescriptionViewModel
+
+@{
+  ViewData["Title"] = $"Edit description for {Model.GroupLabel} group (optional)";
+  ViewData["Application"] = "Tracking System";
+  ViewData["HeaderPath"] = $"{Configuration["AppRootPath"]}/TrackingSystem/Centre/Dashboard";
+  ViewData["HeaderPathName"] = "Tracking System";
+}
+
+<link rel="stylesheet" href="@Url.Content("~/css/trackingSystem/delegates/groupDelegates.css")" asp-append-version="true">
```
[the remark this answers is not in the document]
Comment (showkath10) [reply DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/DelegateGroups/EditDelegateGroupDescription.cshtml]: Sorry, I thought all delegate group related views may use this style sheet but yes you are right we are not using it for this view so I removed this

File: DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/DelegateGroups/EditDelegateGroupDescription.cshtml [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+
+@section NavMenuItems {
+  <partial name="~/Views/TrackingSystem/Shared/_NavMenuItems.cshtml" />
+}
+
+<div class="nhsuk-grid-row">
+  <div class="nhsuk-grid-column-full">
+    <h1 class="nhsuk-heading-l">@ViewData["Title"]</h1>
+    <div class="nhsuk-grid-column-two-thirds">
```
[the remark this answers is not in the document]
Comment (showkath10) [reply DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/DelegateGroups/EditDelegateGroupDescription.cshtml]: I have removed it. yes it was used to align text box different width but now it's removed as suggested.

File: DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/DelegateGroupsController.cs [hunk 15]
```diff
@@ -179,6 +179,38 @@ public IActionResult GroupCourses(int groupId, int page = 1)
...
+            }
+
+            var model = new EditDelegateGroupDescriptionViewModel(group);
+            return View(model);
+        }
+
+        [Route("{groupId:int}/EditDelegateGroupDescription")]
+        [HttpPost]
+        public IActionResult EditDelegateGroupDescription(EditDelegateGroupDescriptionViewModel model, int groupId)
+        {
```
[the remark this answers is not in the document]
Comment (showkath10) [reply DigitalLearningSolutions.Web/Controllers/TrackingSystem/Delegates/DelegateGroupsController.cs:202]: Sorry, I missed it. Added the model validation now

File: DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/Delegates/DelegateGroupsControllerTests.cs [hunk 16]
```diff
@@ -316,5 +316,39 @@ public void GroupDelegatesRemove_should_call_remove_progress_if_checked()
...
+            var model = new EditDelegateGroupDescriptionViewModel();
+            A.CallTo(() => groupsDataService.UpdateGroupDescription(A<int>._, A<int>._, A<string>._))
+                .Returns(false);
+
+            // When
+            var result = delegateGroupsController.EditDelegateGroupDescription(model, 6);
+
+            // Them
+            A.CallTo(() => groupsDataService.UpdateGroupDescription(A<int>._, A<int>._, A<string>._))
+                .MustHaveHappened();
```
[the remark this answers is not in the document]
Comment (showkath10) [reply DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/Delegates/DelegateGroupsControllerTests.cs]: I removed the above assertion

File: DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/Delegates/DelegateGroupsControllerTests.cs [hunk 17]
```diff
@@ -316,5 +316,39 @@ public void GroupDelegatesRemove_should_call_remove_progress_if_checked()
...
+            var model = new EditDelegateGroupDescriptionViewModel();
+            A.CallTo(() => groupsDataService.UpdateGroupDescription(A<int>._, A<int>._, A<string>._))
+                .Returns(true);
+
+            // When
+            var result = delegateGroupsController.EditDelegateGroupDescription(model, 6);
+
+            // Them
+            A.CallTo(() => groupsDataService.UpdateGroupDescription(A<int>._, A<int>._, A<string>._))
+                .MustHaveHappened();
```
[the remark this answers is not in the document]
Comment (showkath10) [reply DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/Delegates/DelegateGroupsControllerTests.cs]: I removed the above assertion

File: DigitalLearningSolutions.Data/DataServices/GroupsDataService.cs (hunk 1 above)
Comment (showkath10) [reply DigitalLearningSolutions.Data/DataServices/GroupsDataService.cs]: I updated method name to IsGroupDescriptionUpdated

File: DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/DelegateGroups/EditDelegateGroupDescription.cshtml [hunk 18]
```diff
@@ -0,0 +1,40 @@
+﻿@inject IConfiguration Configuration
+@using DigitalLearningSolutions.Web.ViewModels.TrackingSystem.Delegates.DelegateGroups
+@using Microsoft.Extensions.Configuration
+@model EditDelegateGroupDescriptionViewModel
+
+@{
+  ViewData["Title"] = $"Edit description for {Model.GroupLabel} group (optional)";
+  ViewData["Application"] = "Tracking System";
+  ViewData["HeaderPath"] = $"{Configuration["AppRootPath"]}/TrackingSystem/Centre/Dashboard";
+  ViewData["HeaderPathName"] = "Tracking System";
```
[the remark this answers is not in the document]
Comment (showkath10) [reply DigitalLearningSolutions.Web/Views/TrackingSystem/Delegates/DelegateGroups/EditDelegateGroupDescription.cshtml]: I have updated as suggested

File: DigitalLearningSolutions.Web/ViewModels/TrackingSystem/Delegates/DelegateGroups/EditDelegateGroupDescriptionViewModel.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Web/ViewModels/TrackingSystem/Delegates/DelegateGroups/EditDelegateGroupDescriptionViewModel.cs]: I'm not sure you mean the AdminUser? I don't see why that would have a GroupLabel. I suspect you mean Group and I agree that sometimes consistency is good, but I think it's usually worth converting the database name to something more clear when converting to the view model. Particularly with some of the really poorly named columns we have on some tables

File: DigitalLearningSolutions.Data/DataServices/GroupsDataService.cs (hunk 1 above)
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data/DataServices/GroupsDataService.cs]: I don't think IsGroupDescriptionUpdated is the correct name here, that suggests the method is checking if the group name has been updated or not rather than it is going to do an update. TryUpdateGroupDescription is perhaps closer to what we need, following a similar convention to TryParse.

File: DigitalLearningSolutions.Data/DataServices/GroupsDataService.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data/DataServices/GroupsDataService.cs]: The method still doesn't fully populate the Group object, look at GetGroupsForCentre for a SQL query to base this on.

File: DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs [hunk 19]
```diff
@@ -1,294 +1,344 @@
...
+
+                //Then
+                var result = groupsDataService.GetGroupAtCentreById(groupId, incorrectCentreId);
+                result?.GroupDescription.Should().NotBe(expectedDescription);
+            }
+            finally
+            {
+                transaction.Dispose();
+            }
+        }
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs:260]: The usage of the transactions is now correct, however you haven't addressed the fact that we aren't testing the correct thing here. The test method name also needs updating to match the new method name.

This test should have 2 assertions:

1. checking the return type of the method (false for this test)
2. the value in the database has NOT been updated. To do this we will need a helper method to query the actual value based on only the groupId. 

You will also need to fix the database value on test that got updated by the previous version of these tests without the transaction (it should be NULL) otherwise we won't be able to tell that these tests are failing.

**Review by AlexJacksonDS — CHANGES_REQUESTED** [review]
Still a few things missing including most of the stuff from the main comment on the previous review. Copying them here again for reference:

The GroupsDataServiceTests and GroupsDataService have had their line endings changed from CRLF to LF, can you change these back (let me know if you don't know how). Can you check that git config core.autocrlf returns false, and if not set this with git config --global core.autocrlf false. This should prevent line endings changes happening unexpectedly, but if you do come across any files with mixed line ending, then CRLF is the ending we should be using. (We think some of the merge tools could also have been messing this up in the past)

There are no automated accessibility tests for the new page. In this case it should be simple, adding a new InlineData in BasicAuthenticatedAccessibilityTests using a suitable groupId for the centre.

There is also a missed requirement on the ticket, ~~setting the maximum length of the Description field to 1000 characters~~, including **creating a migration to do this on the database.**

File: DigitalLearningSolutions.Data/DataServices/GroupsDataService.cs (hunk 1 above)
Comment (showkath10) [reply DigitalLearningSolutions.Data/DataServices/GroupsDataService.cs]: Method name is udpated to TryUpdateGroupDescription

**Comment by showkath10** [comment]
> Still a few things missing including most of the stuff from the main comment on the previous review. See also the comments I've unresolved above. Copying them here again for reference:
> 
> The GroupsDataServiceTests and GroupsDataService have had their line endings changed from CRLF to LF, can you change these back (let me know if you don't know how). Can you check that git config core.autocrlf returns false, and if not set this with git config --global core.autocrlf false. This should prevent line endings changes happening unexpectedly, but if you do come across any files with mixed line ending, then CRLF is the ending we should be using. (We think some of the merge tools could also have been messing this up in the past)
> 
> There are no automated accessibility tests for the new page. In this case it should be simple, adding a new InlineData in BasicAuthenticatedAccessibilityTests using a suitable groupId for the centre.
> 
> There is also a missed requirement on the ticket, ~setting the maximum length of the Description field to 1000 characters~, including **creating a migration to do this on the database.**

image

Sorry, I can't find LF endings. I also selected CRLF endings in Visual Studio. You still see them?

**Comment by AlexJacksonDS** [comment]
> Sorry, I can't find LF endings. I also selected CRLF endings in Visual Studio. You still see them?

Yeah I still see them for both the files I mentioned. Have you set the autocrlf setting I mentioned?
image
image

File: DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs [hunk 20]
```diff
@@ -1,294 +1,344 @@
...
+        [Test]
+        public void UpdateGroupDescription_with_incorrect_centreId_does_not_update_record()
+        {
+            using var transaction = new TransactionScope(TransactionScopeAsyncFlowOption.Enabled);
+            try
+            {
+                // Given
+                const int incorrectCentreId = 107;
+                const int groupId = 5;
+                const string expectedDescription = "Test group description";
```
[the remark this answers is not in the document]
Comment (showkath10) [reply DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs:247]: They used within try block and also they are consistent with other functions in the same class so would not be required to move outside try block

**Comment by showkath10** [comment]
> > Still a few things missing including most of the stuff from the main comment on the previous review. See also the comments I've unresolved above. Copying them here again for reference:
> > The GroupsDataServiceTests and GroupsDataService have had their line endings changed from CRLF to LF, can you change these back (let me know if you don't know how). Can you check that git config core.autocrlf returns false, and if not set this with git config --global core.autocrlf false. This should prevent line endings changes happening unexpectedly, but if you do come across any files with mixed line ending, then CRLF is the ending we should be using. (We think some of the merge tools could also have been messing this up in the past)
> > There are no automated accessibility tests for the new page. In this case it should be simple, adding a new InlineData in BasicAuthenticatedAccessibilityTests using a suitable groupId for the centre.
> > There is also a missed requirement on the ticket, ~setting the maximum length of the Description field to 1000 characters~, including **creating a migration to do this on the database.**
> 
> image
> 
> Sorry, I can't find LF endings. I also selected CRLF endings in Visual Studio. You still see them?

I have no changed git settings to have CRLF endings.

File: DigitalLearningSolutions.Data/DataServices/GroupsDataService.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (showkath10) [reply DigitalLearningSolutions.Data/DataServices/GroupsDataService.cs]: This has been updated to load full group object

File: DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs (hunk 20 above)
[the remark this answers is not in the document]
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs:247]: Technically they aren't consistent because we have no consistency, but it's not really worth changing. I'm more inclined to remove as many of the thing that can't throw an error in the tests outside the try block assuming it doesn't leave us with split // Given groupings.

File: DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs (hunk 19 above)
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs:260]: The test database should be restored to the state it was before the incorrect test was run, i.e. having the description as NULL. We need to keep it as consistent as we can so that if someone else were to write a set of tests relying on that record we don't get unexpected test errors due to differences in the database.

If you want the value to be there for a test for some reason, it then becomes something you need to write a script for so that everyone else can make the same change to their test databases.

File: DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs [hunk 21]
```diff
@@ -1,294 +1,344 @@
...
+
+        [Test]
+        public void GetGroupName_returns_null_with_incorrect_centre()
+        {
+            // When
+            var result = groupsDataService.GetGroupAtCentreById(5, 1)?.GroupLabel;
+
+            // Then
+            result.Should().BeNull();
+        }
```
[the remark this answers is not in the document]
Comment (showkath10) [reply DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs]: I have added two new tests and removed old ones as suggested

File: DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs (hunk 19 above)
Comment (showkath10) [reply DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs]: Restored test database and tested all test cases. All are passed

File: DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs [hunk 22]
```diff
@@ -211,6 +215,56 @@ public void GetRelatedProgressIdForGroupDelegate_returns_null_with_no_record()
...
+                // Given
+                const int centerId = 101;
+                const int groupId = 5;
+                const string expectedDescription = "Test group description1";
+
+                // When
+                var isUpdated = groupsDataService.TryUpdateGroupDescription(groupId, centerId, expectedDescription);
+
+                // Then
+                isUpdated.Should().Be(true);
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs:233]: There are different methods BeTrue and BeFalse. They do the same thing as Be(true)/Be(false), but can you use those instead for this and the next test

File: DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs [hunk 23]
```diff
@@ -288,7 +343,17 @@ private void AddProgressRecordForGroupWithSharedCourse()
...
                     VALUES (285172,299228,25918,1,GETUTCDATE(),3,0)
-                    SET IDENTITY_INSERT dbo.Progress OFF");
+                    SET IDENTITY_INSERT dbo.Progress OFF"
+            );
+        }
+
+        private string? GetGroupDescriptionById(int groupId)
+        {
+            return connection.Query<string?>(
+                @"SELECT GroupLabel FROM Groups
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs:353]: This is the wrong column, we should be getting GroupDescription (and you'll find that this causes the "doesn't update" unit test to fail unless you revert the changes to the test databases caused by the incorrect tests

File: DigitalLearningSolutions.Data/DataServices/GroupsDataService.cs [hunk 24]
```diff
@@ -218,5 +198,61 @@ public void AddDelegateToGroup(int delegateId, int groupId, DateTime addedDate,
...
+                @"UPDATE Groups
+                    SET
+                        GroupDescription = @groupDescription
+                    WHERE GroupID = @groupId AND CentreId = @centreId",
+                new { groupDescription, groupId, centreId }
+            );
+            return numberOfAffectedRows > 0;
+        }
+
+        private const string CourseCountSql = @"SELECT COUNT(*)
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Data/DataServices/GroupsDataService.cs:247]: This should be at the top of the file, where the formatter puts it.

File: DigitalLearningSolutions.Web.AutomatedUiTests/AccessibilityTests/BasicAuthenticatedAccessibilityTests.cs [hunk 25]
```diff
@@ -60,7 +60,7 @@ public class BasicAuthenticatedAccessibilityTests : AccessibilityTestsBase,
         [InlineData("/NotificationPreferences/Edit/AdminUser", "Update notification preferences")]
         [InlineData("/NotificationPreferences/Edit/DelegateUser", "Update notification preferences")]
         [InlineData("/ChangePassword", "Change password")]
-        [InlineData("/TrackingSystem/Support", "Support")]
+        [InlineData("/TrackingSystem/Delegates/Groups/103/EditDescription", "Edit description for ABC Digital Supervision 2 group (optional)")]
```
Comment (AlexJacksonDS) [inline DigitalLearningSolutions.Web.AutomatedUiTests/AccessibilityTests/BasicAuthenticatedAccessibilityTests.cs:63]: The accessibility test is failing because group 103 is not in the base version of the database, so we have different ones. Use group 5 instead since we're using that for all the other tests and move this next to the other similar tests (lines 46-51).

Also you've removed the Support test, this needs bringing back

**Review by AlexJacksonDS — CHANGES_REQUESTED** [review]
There's just a few final things to tidy up, and a pointer for fixing your failing accessibility test in the jenkins build

**Review by AlexJacksonDS — APPROVED** [review]
Looks OK now, just still need to sort out the test database before this is merged.

File: DigitalLearningSolutions.Data/DataServices/GroupsDataService.cs [hunk 26]
```diff
@@ -218,5 +209,50 @@ public void AddDelegateToGroup(int delegateId, int groupId, DateTime addedDate,
                 new { groupId, delegateId, addedDate, addedByFieldLink }
             );
         }
+
+        public Group? GetGroupAtCentreById(int groupId, int centreId)
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Data/DataServices/GroupsDataService.cs:213]: This method looks to be duplicating the entirety of the GetGroupsAtCentre method, but instead just retrieving a single group. That seems a little overkill to me.

Could we instead make a new GroupService method called GetGroupAtCentreById that calls the old GetGroupsAtCentre and then only returns the group with the supplied groupId?

File: DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs [hunk 27]
```diff
@@ -211,6 +215,56 @@ public void GetRelatedProgressIdForGroupDelegate_returns_null_with_no_record()
             result.Should().BeNull();
         }
 
+        [Test]
+        public void UpdateGroupDescription_updates_record()
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs:219]: Could we change this unit test name to
"TryUpdateGroupDescription_updates_record_with_correct_centreId" 
so that it matches the method name and circumstances.

File: DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/Delegates/DelegateGroupsControllerTests.cs [hunk 28]
```diff
@@ -316,5 +316,35 @@ public void GroupDelegatesRemove_should_call_remove_progress_if_checked()
...
+
+            // When
+            var result = delegateGroupsController.EditDescription(model, 6);
+
+            // Then
+            result.Should().BeNotFoundResult();
+        }
+
+        [Test]
+        public void EditDelegatesGroupDescription_should_redirect_to__index_action()
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Web.Tests/Controllers/TrackingSystem/Delegates/DelegateGroupsControllerTests.cs:336]: Very minor, but I think there are two underscores in-between "to" and "index".

File: DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs [hunk 29]
```diff
@@ -211,6 +215,56 @@ public void GetRelatedProgressIdForGroupDelegate_returns_null_with_no_record()
...
+        [Test]
+        public void TryUpdateGroupDescription_with_incorrect_centreId_does_not_update_record()
+        {
+            using var transaction = new TransactionScope(TransactionScopeAsyncFlowOption.Enabled);
+            try
+            {
+                // Given
+                const int incorrectCentreId = 107;
+                const int groupId = 5;
+                const string expectedDescription = "Test group description";
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs:252]: Could we rename this "newDescription"?

File: DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs [hunk 30]
```diff
@@ -211,6 +215,56 @@ public void GetRelatedProgressIdForGroupDelegate_returns_null_with_no_record()
...
+
+                // When
+                var isUpdated = groupsDataService.TryUpdateGroupDescription(groupId, incorrectCentreId, expectedDescription);
+
+                //Then
+                isUpdated.Should().Be(false);
+                var result = this.GetGroupDescriptionById(groupId);
+                result?.Should().NotBe(expectedDescription);
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs:260]: Can we make this line 
`result.Should().BeNull();`
so then we are explicitly testing that the value is the same as how it was.

File: DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs [hunk 31]
```diff
@@ -211,6 +215,56 @@ public void GetRelatedProgressIdForGroupDelegate_returns_null_with_no_record()
...
+                const int incorrectCentreId = 107;
+                const int groupId = 5;
+                const string expectedDescription = "Test group description";
+
+                // When
+                var isUpdated = groupsDataService.TryUpdateGroupDescription(groupId, incorrectCentreId, expectedDescription);
+
+                //Then
+                isUpdated.Should().Be(false);
+                var result = this.GetGroupDescriptionById(groupId);
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Data.Tests/DataServices/GroupsDataServiceTests.cs:259]: "this." is redundant here.

**Review by DanBloxham-sw — CHANGES_REQUESTED** [review]
A number of minor changes to make.

**Comment by showkath10** [comment]
New MR is created so this is no longer needed.
New MR - https://github.com/TechnologyEnhancedLearning/DLSV2/pull/706
